File: src/AutoRest.SdkExplorer.TypeScript/src/Model/Example/ExampleDesc.ts
import { convertStringIndexArrayFromMap, convertStringIndexdArray, convertStringIndexdArrayToMap, getStringIndexArrayFromMap } from "../../Utils/utils";
import { ExampleValueDesc } from "./ExampleValueDesc";

export class ExampleDesc implements AutoRest.SdkExplorer.Interface.ExampleDesc {
    language: string;
    sdkPackageName: string;
    sdkPackageVersion: string;
    explorerCodeGenVersion: string;
    generatedTimestamp: string;
    serviceName: string;
    resourceName: string;
    operationName: string;
    swaggerOperationId: string;
    sdkOperationId: string;
    exampleName: string;
    originalFilePath: string;
    originalFileNameWithoutExtension: string;
    embeddingText: string;
    embeddingVector: string;

    get exampleValues(): { [index: string]: ExampleValueDesc } | undefined {
        return getStringIndexArrayFromMap(this.exampleValuesMap);
    }

    exampleValuesMap?: Map<string, ExampleValueDesc>;

    constructor(data: AutoRest.SdkExplorer.Interface.ExampleDesc) {
        this.language = data.language!;
        this.sdkPackageName = data.sdkPackageName!;
        this.sdkPackageVersion = data.sdkPackageVersion!;
        this.explorerCodeGenVersion = data.explorerCodeGenVersion!;
        this.generatedTimestamp = data.generatedTimestamp!;
        this.serviceName = data.serviceName!;
        this.resourceName = data.resourceName!;
        this.operationName = data.operationName!;
        this.swaggerOperationId = data.swaggerOperationId!;
        this.sdkOperationId = data.sdkOperationId!;
        this.exampleName = data.exampleName!;
        this.originalFilePath = data.originalFilePath!;
        this.originalFileNameWithoutExtension = data.originalFileNameWithoutExtension!;
        this.embeddingText = data.embeddingText!;
        this.embeddingVector = data.embeddingVector!;

        this.exampleValuesMap = convertStringIndexdArrayToMap(data.exampleValues, (d, k) => new ExampleValueDesc(d, this, k));
    }

    public setExampleValues(values: Map<string, ExampleValueDesc>) {
        this.exampleValuesMap = new Map<string, ExampleValueDesc>();
        values.forEach((v, k) => {
            this.exampleValuesMap!.set(k, v);
        })
    }

    public toPayload(): AutoRest.SdkExplorer.Interface.ExampleDesc {
        const r: AutoRest.SdkExplorer.Interface.ExampleDesc = {
            language: this.language,
            sdkPackageName: this.sdkPackageName,
            sdkPackageVersion: this.sdkPackageVersion,
            explorerCodeGenVersion: this.explorerCodeGenVersion,
            generatedTimestamp: this.generatedTimestamp,
            serviceName: this.serviceName,
            resourceName: this.resourceName,
            operationName: this.operationName,
            swaggerOperationId: this.swaggerOperationId,
            sdkOperationId: this.sdkOperationId,
            exampleName: this.exampleName,
            originalFilePath: this.originalFilePath,
            originalFileNameWithoutExtension: this.originalFileNameWithoutExtension,
            embeddingText: this.embeddingText,
            embeddingVector: this.embeddingVector,
            exampleValues: this.exampleValuesMap === undefined ? undefined : convertStringIndexArrayFromMap(this.exampleValuesMap, (v, k) => v.toPayload())
        };
        return r;
    }

    public toJson(): string {
        return JSON.stringify(this.toPayload(), (key: string, value: any) => {
            const toIgnore = [
                "serializerName",
                "modelName",
            ];
            if (toIgnore.includes(key))
                return undefined;
            return value;
        }, "  ");
    }
}

File: src/AutoRest.SdkExplorer/Model/Example/ExampleDesc.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


using System;
using System.Collections.Generic;
using AutoRest.SdkExplorer.Utilities;

namespace AutoRest.SdkExplorer.Model.Example
{
    public class ExampleDesc
    {
        public string? Language { get; set; }
        public string? SdkPackageName { get; set; }
        public string? SdkPackageVersion { get; set; }
        public string? ExplorerCodeGenVersion { get; set; }
        public string? GeneratedTimestamp { get; set; }

        public string? ServiceName { get; set; }
        public string? ResourceName { get; set; }
        public string? OperationName { get; set; }
        public string? SwaggerOperationId { get; set; }
        public string? SdkOperationId { get; set; }
        public string? SdkFullUniqueName { get; set; }

        public string? ExampleName { get; set; }
        public string? OriginalFilePath { get; set; }
        public string? OriginalFileNameWithoutExtension { get; set; }

        public string? AiDescription { get; set; }

        public string? EmbeddingText { get; set; }
        public string? EmbeddingVector { get; set; }

        public Dictionary<string, ExampleValueDesc> ExampleValues { get; set; } = new Dictionary<string, ExampleValueDesc>();

        public ExampleDesc()
        {

        }

        public string ToYaml()
        {
            return this.SerializeToYaml();
        }
    }
}
